Add "disable taps during gameplay" to touch input settings and related UI

Susko3 opened the pull request:
- Closes https://github.com/ppy/osu/issues/25065

Adds a new setting (`OsuSetting.GameplayDisableTaps`) and changes existing strings to be in line with https://github.com/ppy/osu/issues/25065#issuecomment-1786360644.

| Area | Old string | New string | Notes |
|--------|--------|--------|--------|
| `TouchSettings` for all platforms | N/A | "Disable taps during gameplay" | Newly added |
| beatmap loading screen `InputSettings` | "Disable mouse buttons during gameplay"  | "Disable clicks during gameplay" or "Disable taps during gameplay" | currently based on `RuntimeInfo.IsDesktop`, but can be changed to use https://github.com/ppy/osu/pull/25348|
| SDL2 and Android `MouseSettings` | "Disable mouse buttons during gameplay" | "Disable clicks during gameplay" | |

Susko3 committed:
Add disable taps checkbox to touch input settings

Susko3 committed:
Rename mouse button string to `Disable clicks during gameplay`

Susko3 committed:
Rename popup/binding string to `Toggle gameplay clicks/taps`

Susko3 committed:
Use appropriate clicks/taps text in player loader input settings

peppy commented:
I don't understand. Should this not be adding a new tracked setting? It feels so wrong to have both toggles affect the same underlying setting. What's the point?

Susko3 commented:
> It feels so wrong to have both toggles affect the same underlying setting. What's the point?

I thought we were going for option 1 in https://github.com/ppy/osu/issues/25065#issuecomment-1784593467. But I can agree with adding a new setting that affects only touch (and changing the mouse setting to only affect mouse). (I haven't considered adding a new underlying setting up until now.)

> Should this not be adding a new tracked setting?

For now, I'm against adding a new hotkey (and a tracked settings to go with it).

peppy commented:
By tracked setting I just meant a new `OsuConfigManager` entry. I agree it shouldn't be OSD tracked.

Susko3 committed:
Add new setting for GameplayDisableTaps

Susko3 committed:
Make tests fail

Susko3 committed:
Fix tests by using the correct setting for touch input

Susko3 committed:
Update UI to use the new setting

Susko3 committed:
Revert "Rename popup/binding string to `Toggle gameplay clicks/taps`"

This reverts commit 0d8bfedf5d3693809e76471b9feb47ebf140e40b.

Susko3 commented on osu.Game/Rulesets/UI/RulesetInputManager.cs:
```diff
@@ -147,9 +149,9 @@ protected override bool Handle(UIEvent e)
 
         protected override bool HandleMouseTouchStateChange(TouchStateChangeEvent e)
         {
-            if (mouseDisabled.Value)
+            if (tapsDisabled.Value)
             {
-                // Only propagate positional data when mouse buttons are disabled.
+                // Only propagate positional data when taps are disabled.
                 e = new TouchStateChangeEvent(e.State, e.Input, e.Touch, false, e.LastPosition);
             }
```
This isn't doing much. All the rulesets now support touch natively and don't need mouse mouse emulation. And therefore completely ignore this input path. Possibly still relevant for custom rulesets.

frenzibyte replied:
This path is specific to the osu! ruleset, and it should be removed now that osu! handles touch input appropriately with the `OsuTouchInputMapper` component (as well as correctly ignoring taps while still maintaining position, without doing it weirdly like in the code highlighted here).

But that should be done in a separate PR to keep review complexity to a minimum.

peppy replied:
> This path is specific to the osu! ruleset

How so? I don't understand. This should apply to any ruleset which supports mouse but not explicitly touch?

frenzibyte replied:
Hmm, I guess freeform custom rulesets would end up with a non-functioning setting that way...fair enough.

frenzibyte commented on osu.Game/Configuration/OsuConfigManager.cs:
```diff
@@ -412,6 +414,7 @@ public enum OsuSetting
         AutomaticallyDownloadMissingBeatmaps,
-        EditorShowSpeedChanges
+        EditorShowSpeedChanges,
+        GameplayDisableTaps,
```
I would want to include `Touch` anywhere in this setting name so I can easily find it, "Gameplay" sounds like a broad term to me for what is meant to be a touch-specific setting.

peppy replied:
`TouchDisableGameplayTaps`

frenzibyte commented on osu.Game/Overlays/Settings/Sections/Input/TouchSettings.cs:
```diff
@@ -3,38 +3,48 @@
+        private void load(OsuConfigManager osuConfig)
         {
-            Children = new Drawable[]
+            if (!RuntimeInfo.IsMobile) // don't allow disabling the only input method (touch) on mobile.
```
I would avoid including changes irrelevant from the PR's target, but I guess this one may be fine.

peppy replied:
I'd still prefer it to be separate, for better changelog tracking.

Susko3 replied:
I can move this and the `OsuGameAndroid` change to a new pull titled "Show touch settings on Android" if you wish.

peppy replied:
Yep, sounds good.

peppy commented on osu.Game/Screens/Play/PlayerSettings/InputSettings.cs:
```diff
@@ -19,12 +20,13 @@ public InputSettings()
             {
                 mouseButtonsCheckbox = new PlayerCheckbox
```
I'd rename the variable to `disableGameplayInputCheckbox`.

Susko3 replied:
👍 
Missed that one in the refactor. But I think I'll just move this to BDL and scrap the variable entirely.

Susko3 committed:
Rename setting to `TouchDisableGameplayTaps` for better visibility when searching

Susko3 committed:
Move `InputSettings` children creation code to BDL

- Avoids now obsolete variable name
- Makes changing to touch detection easier (access to session statics in BDL)

Susko3 committed:
Remove mobile specific changes

Will be added back in a separate PR

Susko3 committed:
Remove unused using

peppy committed:
Remove unused using statements

peppy committed:
Merge branch 'master' into better-touch-settings

peppy commented on osu.Game/Screens/Play/PlayerSettings/InputSettings.cs:
```diff
@@ -23,9 +23,8 @@ private void load(OsuConfigManager config)
             {
                 new PlayerCheckbox
                 {
-                    // TODO: change to touchscreen detection once https://github.com/ppy/osu/pull/25348 makes it in
```
I felt like this change made sense, but I guess it's not a big deal either way for now.
